src/components/Export.tsx

```diff
@@ -0,0 +1,358 @@
+  stopRecording: () => void;
+  defaultImagePrefix?: string;
+  disabled?: boolean;
+};
+
+export const TEST_ID_EXPORT_ACTION_BUTTON = "export-action";
+export const TEST_ID_OPEN_EXPORT_MODAL_BUTTON = "open-export-modal";
+
+const defaultProps: Partial<ExportButtonProps> = {
+  defaultImagePrefix: "image",
+  disabled: false,
+};
+
+const HorizontalDiv = styled.div`
+  display: flex;
+  flex-direction: row;
+  gap: 6px;
+  flex-wrap: wrap;
+`;
+
+const VerticalDiv = styled.div`
+  display: flex;
+  flex-direction: column;
+  gap: 10px;
+`;
+
+const CustomRangeDiv = styled(HorizontalDiv)`
+  & input {
+    width: 70px;
+    text-align: right;
+  }
+`;
+
+function clamp(value: number, min: number, max: number): number {
+  return Math.min(max, Math.max(value, min));
+}
+
+/**
+ * A single Export button that opens up an export modal when clicked. Manages starting and stopping
+ * an image sequence recording, resetting state when complete.
+ */
+export default function Export(inputProps: ExportButtonProps): ReactElement {
+  const props = { ...defaultProps, ...inputProps } as Required<ExportButtonProps>;
+
+  const theme = useContext(AppThemeContext);
+
+  const enum ExportMode {
+    ALL,
+    CURRENT,
+    CUSTOM,
+  }
+
+  // Static convenience method for creating simple modals + notifications.
+  // Used here for the cancel modal and the success notification.
+  // Note: notification API seems to only place notifications at the top-level under the
+  // <body> tag, which causes some issues with styling.
+  const { modal, notification } = App.useApp();
+  const modalContextRef = useRef<HTMLDivElement>(null);
+
+  const [isLoadModalOpen, setIsLoadModalOpen] = useState(false);
+  const [isRecording, setIsRecording] = useState(false);
+  const [isPlayingCloseAnimation, setIsPlayingCloseAnimation] = useState(false);
+
+  const [exportMode, setExportMode] = useState(ExportMode.ALL);
+  const [customMin, setCustomMin] = useState(0);
+  const [customMax, setCustomMax] = useState(props.totalFrames - 1);
+  const [imagePrefix, setImagePrefix] = useState(props.defaultImagePrefix);
+  const [useDefaultImagePrefix, setUseDefaultImagePrefix] = useState(true);
+  const [frameIncrement, setFrameIncrement] = useState(1);
+
+  const [originalFrame, setOriginalFrame] = useState(props.currentFrame);
+  const [percentComplete, setPercentComplete] = useState(0);
+
+  // If dataset changes, update the max range field with the total frames.
+  useEffect(() => {
+    setCustomMax(props.totalFrames - 1);
+  }, [props.totalFrames]);
+
+  useEffect(() => {
+    if (useDefaultImagePrefix) {
+      setImagePrefix(props.defaultImagePrefix);
+    }
+  }, [props.defaultImagePrefix, useDefaultImagePrefix]);
+
+  // Store the current frame whenever the modal opens so we can reset to it when
+  // the modal is closed.
+  useEffect(() => {
+    if (isLoadModalOpen) {
+      setOriginalFrame(props.currentFrame);
+    }
+  }, [isLoadModalOpen]);
```

Comment:
Again setting state in effects, though here I'd recommend replacing this with a ref. A rule of thumb I've started thinking about for state is: if somehow this value changed for no reason, would I expect the appearance of the component to change? If the answer is "yes" or even "maybe," it belongs in state. If not, a ref is a fine place for it.

In place of this effect, I'd change `originalFrame` to a ref and set it in the `isLoadModalOpen` setter:

```typescript
const originalFrameRef = useRef(props.currentFrame);
const [isLoadModalOpen, _setIsLoadModalOpen] = useState(false);
const setIsLoadModalOpen = (isOpen) => {
  _setIsLoadModalOpen(isOpen);
  if (isOpen) {
    originalFrameRef.current = props.currentFrame;
  }
}
```

As a bonus, you could then remove the value from the dependency list of `stopRecording` below, since the latest value of the ref will always be accessible there.

Reply:
Thanks for the code example here too!